26-dracula-ant/technical_specification.md

```diff
@@ -0,0 +1,103 @@
+
+## Implementation Details:
+
+### REST API
+
+The Work Package Service REST API should have the following end points:
+
+- `POST /work_packages`
+  - auth header: internal access token
+  - request body:
+    - `dataset_id`: str
+    - `type`: enum (download/upload)
+    - `file_ids`: array of strings  (empty = all files of the dataset)
+    - `public_key`: string (the user's public Crypt4GH key)
+  - response body:
+    - `id`: the work package id
+    - `token`: work package access token
+- `GET /work_packages`
+  - auth header: internal access token
+  - gets all list of all work packages of the current user
+- `GET /work_packages/{id}`
+  - auth header: internal access token
+  - gets work package with given id if it belongs to the current user
+- `DELETE /work_packages/{id}`
+  - auth header: internal access token
+  - deletes work package with given id if it belongs to the current user
+- `GET /work_package`
+  - auth header: work package access token
+  - gets work package that has the same hash as the given token
+    (note that this is not really RESTful, the work package is passed in directly via the token)
+- `GET /work_order_token`
+  - auth header: work package access token
+  - query path parameter: `file_id`
+  - gets an encrypted work order token for the given work package access token
+
+### Work package access token
+
+The work package access token is a random string generated by the Work Package Service on a POST request. Its SHA256 hash is stored in the database. The token itself is not stored, but it is encrypted with the `public_key` and then returned in the response.
+
+### Work order token
+
+This is a JWT with a very short time to live (maximum 30 seconds). It is signed by the Work Package Service and then encrypted with the `public_key` of the user, similar ot the work package access token. It contains the following claims:
+
+- `type`: `download` or `upload`
+- `file_id`: the id of the file that shall be downloaded or uploaded
+- `user_id`: the internal ID of the user
+- `public_key`: the public key stored in the work package
+
+The work package access and work order tokens are both signed by the Work Package Service. A separate key pair (different from the keys used for the internal access tokens) should be created for this purpose. The public key must be also communicated to the download/upload controller services so that they can validate the work order token.
+
+### Dataset access checks
+
+In order to check whether a given user has access to a given dataset, the claims repository provides an *internal* RPC endpoint that would return `false` or `true` according to whether the user has access or not.
+- `/rpc/has_access?dataset_id=...&user_id=...`
+
+This endpoint will be consulted by the Work Package Service as well as the download/upload services controllers.
+
+Access permissions may later be stored in an internal database populated by evens sent by a data access service.
```

Comment:
I would rather propose that the service will receive this information from the visa issuer later.

Reply:
Yes, removed that.